include/boost/geometry/views/detail/boundary_view/implementation.hpp

```diff
@@ -0,0 +1,411 @@
+                                  > const& other) const
+    {
+        return static_cast<Difference>(other.m_index)
+            - static_cast<Difference>(m_index);
+    }
+
+    inline void advance(Difference n)
+    {
+        m_index += n;
+    }
+
+private:
+    Polygon* m_polygon;
+    size_type m_index;
+};
+
+
+template <typename Ring>
+class ring_boundary : closeable_view<Ring, closure<Ring>::value>::type
+{
+private:
+    typedef typename closeable_view<Ring, closure<Ring>::value>::type base_type;
+
+public:
+    typedef typename base_type::iterator iterator;
+    typedef typename base_type::const_iterator const_iterator;
+    
+    typedef linestring_tag tag_type;
+
+    explicit ring_boundary(Ring& ring)
+        : base_type(ring) {}
+
+    iterator begin() { return base_type::begin(); }
+    iterator end() { return base_type::end(); }
+    const_iterator begin() const { return base_type::begin(); }
+    const_iterator end() const { return base_type::end(); }
+};
+
+
+template <typename Geometry, typename Tag = typename tag<Geometry>::type>
+struct views_container_initializer
+{};
+
+template <typename Polygon>
+struct views_container_initializer<Polygon, polygon_tag>
+{
+    typedef polygon_rings_iterator<Polygon> rings_iterator_type;
+
+    static inline std::size_t num_rings(Polygon const& polygon)
+    {
+        return geometry::num_interior_rings(polygon) + 1;
+    }
+
+    template <typename BoundaryView>
+    static inline void apply(Polygon const& polygon, BoundaryView* views)
+    {
+        std::uninitialized_copy(rings_iterator_type(polygon),
+                                rings_iterator_type(polygon, true),
+                                views);
+    }
+};
+
+template <typename MultiPolygon>
+struct views_container_initializer<MultiPolygon, multi_polygon_tag>
+{
+    typedef typename boost::mpl::if_
+        <
+            boost::is_const<MultiPolygon>,
+            typename boost::range_value<MultiPolygon>::type const,
+            typename boost::range_value<MultiPolygon>::type
+        >::type polygon_type;
+
+    typedef polygon_rings_iterator<polygon_type> rings_iterator_type;
+
+    static inline std::size_t num_rings(MultiPolygon const& multipolygon)
+    {
+        return geometry::num_interior_rings(multipolygon)
+            + static_cast<std::size_t>(boost::size(multipolygon));
+    }
+
+    template <typename BoundaryView>
+    static inline void apply(MultiPolygon const& multipolygon,
+                             BoundaryView* views)
+    {
+        BoundaryView* cur_it = views;
+        for (typename boost::range_iterator<MultiPolygon>::type
+                 it = boost::begin(multipolygon);
+             it != boost::end(multipolygon);
+             ++it)
+        {
+            cur_it = std::uninitialized_copy(rings_iterator_type(*it),
+                                             rings_iterator_type(*it, true),
+                                             cur_it);
+        }
```

Comment:
I'm glad that you extracted this to be a separate function, because now it's more clear that this function it is not safe. If one of the calls to `std::uninitialized_copy` fails then the elements before `cur_it` are constructed and the elements after `cur_it` are left not constructed. And the former elements aren't destroyed here (and they should be). You're trying to handle the case in `clear()` but AFAIU this function tries to call the destructors of ALL of the objects. In this case only a part of them is created so the destructors are called for not created objects.